### PR TITLE
Feat 351 action read endpoint

### PR DESCRIPTION
**Pull request by opyate**
Action read endpoint, with a few extras:
- hooked in extra pre-commit checks
- refactored the model/schema/crud a bit

### CODE REVIEW

File: loader/pyproject.toml [hunk 1]
```diff
@@ -16,6 +16,9 @@ csvkit = "^1.0.6"
 pytest = "^7.0.1"
 pydocstyle = { version = "^6.1.1", extras = ["pydocstyle[toml]"] }
 pre-commit = "^2.17.0"
+flake8 = "^4.0.1"
+black = "^22.1.0"
+mypy = "^0.931"
```
Comment (joel-wright) [inline loader/pyproject.toml:21]: thank you!

**Review by joel-wright — COMMENTED** [review]
I have a few questions, but the changes look completely reasonable (and were far smaller than they initially appeared to be!) :)

File: backend/app/api/api_v1/routers/actions.py [hunk 2]
```diff
@@ -1,31 +1,45 @@
...
+from app.db.crud.action import create_action, get_actions_query, is_action_exists
+from app.db.crud.document import create_document
+from app.db.models.document import DocumentInvalidReason
+from app.db.schemas.action import ActionCreate, ActionInDB
+from app.db.schemas.document import DocumentCreateInternal
 from app.db.session import get_db
-from navigator.core.aws import get_s3_client, S3Document
+from fastapi import APIRouter, Depends, HTTPException, Request
+from fastapi_pagination import Page
+from fastapi_pagination.ext.sqlalchemy import paginate
```
[the remark this answers is not in the document]
Comment (opyate) [reply backend/app/api/api_v1/routers/actions.py:16]: I dropped `isort` as imports formatting is included with `black`.

File: backend/app/db/schemas/action.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    def set_date(cls, val):  # noqa: D102
+        return val or 1
+
+
+class Action(_ActionInDBBase):
+    """An Action to return to the client"""
+
+
+class ActionInDB(_ActionInDBBase):
+    """An Action as stored in the DB"""
```
[the remark this answers is not in the document]
Comment (opyate) [reply backend/app/db/schemas/action.py:59]: `_ActionBase` is common to all schemas, but doesn't necessarily contain everything we'd want to return to the caller. We'd typically return a row as it is in the DB.
I've created 2 separate classes, as these will get finer-grained with roles: `ActionInDB` will always be that - an action as it is in the DB. But `Action` will get a companion, perhaps something like `ActionForAdmins`, where they can see extra rows. A better example is probably `Document` - the `is_valid` stuff will only be shown to admins.

File: backend/app/db/models/document.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+
+class DocumentInvalidReason(enum.Enum):
+    """Reasons why a document might be invalid."""
+
+    unsupported_content_type = "unsupported_content_type"
+    net_ssl_error = "net_ssl_error"
+    net_read_error = "net_read_error"
+    net_connection_error = "net_connection_error"
+    net_too_many_redirects = "net_too_many_redirects"
```
[the remark this answers is not in the document]
Comment (opyate) [reply backend/app/db/models/document.py:17]: As a [Postgres enum](https://github.com/climatepolicyradar/navigator/blob/dev/backend%2Falembic%2Fversions%2F0008_add_enum_to_document_to_represent_.py).

File: backend/app/conftest.py [hunk 5]
```diff
@@ -60,16 +59,17 @@ def create_test_db():
...
-    yield
-
-    # Drop the test database
-    drop_database(test_db_url)
+    try:
+        # Run the tests
+        yield
+    finally:
+        # Drop the test database
+        drop_database(test_db_url)
```
[the remark this answers is not in the document]
Comment (opyate) [reply backend/app/conftest.py:73]: Good catch! Try block expanded.

File: backend/app/api/api_v1/routers/actions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joel-wright) [reply backend/app/api/api_v1/routers/actions.py:16]: Honestly I think `isort` is complementary to `black` & gives a nice predictable ordering/grouping of imports. Absolutely doesn't matter in terms of getting this merge done though.

File: backend/app/db/models/document.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joel-wright) [reply backend/app/db/models/document.py:17]: That's not really what I meant. I was asking why we're ultimately storing invalid document entries in the database & what we intend to do with them.

File: backend/app/db/models/__init__.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        sa.INTEGER(), primary_key=True, autoincrement=True, nullable=False
+    )
+    metadata_value_id = sa.Column(
+        sa.Integer, sa.ForeignKey(MetadataValue.metadata_value_id), nullable=False
+    )
+
+    keyword = sa.Column(sa.VARCHAR(length=255), autoincrement=False, nullable=False)
+
+
+class Source(Base):  # noqa: D101
```
[the remark this answers is not in the document]
Comment (opyate) [reply backend/app/db/models/__init__.py:80]: CCLW is the only source for alpha. Future will be extras like [CPD](https://climatepolicydatabase.org/), etc.

Added docstring.

**Review by joel-wright — APPROVED** [review]
LGTM

**Pull request merged**